servers/quarkus-server/src/main/java/org/projectnessie/server/authn/NessieHttpAuthenticator.java

```diff
@@ -0,0 +1,105 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.server.authn;
+
+import io.quarkus.security.AuthenticationFailedException;
+import io.quarkus.security.identity.IdentityProvider;
+import io.quarkus.security.identity.IdentityProviderManager;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.identity.request.AnonymousAuthenticationRequest;
+import io.quarkus.vertx.http.runtime.security.HttpAuthenticationMechanism;
+import io.quarkus.vertx.http.runtime.security.HttpAuthenticator;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.ext.web.RoutingContext;
+import javax.annotation.Priority;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Alternative;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+/**
+ * A custom {@link HttpAuthenticator}. This authenticator that performs the following main duties:
+ *
+ * <ul>
+ *   <li>Prevents the Quarkus OIDC authentication mechanism from attempting authentication when it
+ *       is not configured.
+ *   <li>Completely disallows unauthenticated requests when authentication is enabled.
+ * </ul>
+ */
+@Alternative // @Alternative + @Priority ensure the original HttpAuthenticator bean is not used
+@Priority(1)
+@ApplicationScoped
+public class NessieHttpAuthenticator extends HttpAuthenticator {
+
+  @Inject IdentityProviderManager identityProvider;
+
+  @ConfigProperty(name = "nessie.auth.enabled")
+  boolean authEnabled;
+
+  @Inject
+  public NessieHttpAuthenticator(
+      Instance<HttpAuthenticationMechanism> instance, Instance<IdentityProvider<?>> providers) {
+    super(instance, providers);
+  }
+
+  @Override
+  public Uni<SecurityIdentity> attemptAuthentication(RoutingContext context) {
+    if (!authEnabled) {
+      String authHeader = context.request().getHeader(HttpHeaders.AUTHORIZATION.toString());
+      if (authHeader != null) {
+        // Avoid attempting header-based authentication because the server might not have all the
+        // required configuration settings in this case.
+        // Note: The OIDC Quarkus extension has to be enabled at build time, but we cannot provide
+        // working configuration defaults at build time. Therefore, if OIDC were to attempt
+        // authenticating a client request with default config settings it would almost always
+        // produce a 500 error. Since authentication is not enabled in this case, we immediately
+        // return an anonymous SecurityIdentity here.
+        return anonymous();
+      }
+    }
+
+    return super.attemptAuthentication(context)
+        .onItem()
+        .transformToUni(
+            securityIdentity -> {
+              if (securityIdentity == null) {
+                if (authEnabled) {
```

Comment:
How can `authEnabled` be `false` here?

Reply:
it can happen in test. The top-level `if` for the "anonymous" case catches only header-based auth, but some tests use annotation-based auth that does not use the header.

Reply:
AH, okay. Can you add a comment explaining that?

Reply:
I've refactored the configs and `NessieHttpAuthenticator` `if` statements in 9afbec6a279d242e5565ed38bd53523f6ca9ab67. I hope it is clearer now.